apps/vaults/hooks/useSolverPortals.ts

```diff
@@ -0,0 +1,327 @@
+		}
+		return toNormalizedBN(latestQuote?.current?.minBuyAmount, request?.current?.outputToken?.decimals || 18);
+	}, [latestQuote, request]);
+
+	/* 🔵 - Yearn Finance ******************************************************
+	** Retrieve the current outValue from the quote, which will be used to
+	** display the current value to the user.
+	**************************************************************************/
+	const onRetrieveExpectedOut = useCallback(async (request: TInitSolverArgs): Promise<TNormalizedBN> => {
+		if (isSolverDisabled[Solver.PORTALS]) {
+			return toNormalizedBN(0);
+		}
+		const quoteResult = await getQuote(request, true);
+		return toNormalizedBN(formatBN(quoteResult?.minBuyAmount), request.outputToken.decimals);
+	}, [getQuote]);
+
+	/* 🔵 - Yearn Finance ******************************************************
+	** Retrieve the allowance for the token to be used by the solver. This will
+	** be used to determine if the user should approve the token or not.
+	**************************************************************************/
+	const onRetrieveAllowance = useCallback(async (): Promise<TNormalizedBN> => {
+		if (!latestQuote?.current || !request?.current || isSolverDisabled[Solver.PORTALS]) {
+			return toNormalizedBN(0);
+		}
+
+		try {			
+			const approval = await getApproval({
+				network: 1,
+				params: {
+					takerAddress: toAddress(request.current.from),
+					sellToken: toAddress(request.current.inputToken.value),
+					sellAmount:formatBN(request.current.inputAmount || 0).toString(),
+					buyToken: toAddress(request.current.outputToken.value)
+				}
+			});
+
+			if (!approval) {
+				throw new Error('Fail to get approval');
+			}
+
+			return toNormalizedBN(approval.context.allowance, request.current.inputToken.decimals);
+		} catch (error) {
+			console.error(error);
+			return ({raw: Zero, normalized: 0});
+		}
+
+	}, [getApproval]);
+
+	/* 🔵 - Yearn Finance ******************************************************
+	** Trigger an signature to approve the token to be used by the Portals
+	** solver. A single signature is required, which will allow the spending
+	** of the token by the Portals solver.
+	**************************************************************************/
+	const onApprove = useCallback(async (
+		amount = ethers.constants.MaxUint256,
+		txStatusSetter: React.Dispatch<React.SetStateAction<TTxStatus>>,
+		onSuccess: () => Promise<void>
+	): Promise<void> => {
+		if (!latestQuote?.current || !request?.current || isSolverDisabled[Solver.PORTALS]) {
+			return;
+		}
+
+		try {
+			const approval = await getApproval({
+				network: 1,
+				params: {
+					takerAddress: toAddress(request.current.from),
+					sellToken: toAddress(request.current.inputToken.value),
+					sellAmount:formatBN(request.current.inputAmount || 0).toString(),
+					buyToken: toAddress(request.current.outputToken.value)
+				}
+			});
+
+			if (!approval) {
+				throw new Error('Fail to get approval');
+			}
```

Comment:
What this does is just send the code execution to the block of code inside the catch block, it will not make the UI crash because the error is just caught. The benefit of doing it this way is that we can log or toast a better error message, because we know what originated the error